Content/Items/Misc/Weapons.Soilgun.cs

```diff
@@ -0,0 +1,1488 @@
+
+        public override void SetStaticDefaults()
+        {
+            DisplayName.SetDefault("Soil");
+        }
+
+        public override void SetDefaults()
+        {
+            Projectile.DamageType = DamageClass.Ranged;
+            Projectile.penetrate = 1;
+            Projectile.Size = new Vector2(12);
+            Projectile.friendly = true;
+            Projectile.timeLeft = 240;
+        }
+
+        public override bool PreAI()
+        {
+            int VitricSand = Mod.Find<ModItem>("VitricSandItem").Type;
+            if (AmmoType == VitricSand)
+            {
+                if (stuck)
+                {
+                    NPC target = Main.npc[EnemyType];
+                    Projectile.position = target.position + offset;
+                }
+                Projectile.penetrate = -1;
+            }
+            return true;
+        }
+
+        public override void AI()
+        {
+            Time++;
+            Projectile.rotation = Projectile.velocity.ToRotation();
+
+            if (Projectile.timeLeft < 230 && !foundTarget)
+            {
+                Projectile.velocity.Y += 0.96f;
+                if (Projectile.velocity.Y > 16f)
+                    Projectile.velocity.Y = 16f;
+            }
+            if (!Main.dedServ)
+            {
+                ManageCaches();
+                ManageTrail();
+            }
+            #region soil-specific ai
+            int VitricSand = Mod.Find<ModItem>("VitricSandItem").Type;
+            if (AmmoType == VitricSand)
+            {
+                if (stuck)
+                {
+                    if (!Main.npc[EnemyType].active)
+                        Projectile.Kill();
+                    int decreasing = 0;
+                    for (int i = 0; i < Time / 2; i += 10)
+                    {
+                        decreasing += 3;
+                    }
+                    if (Projectile.timeLeft < 120 && Main.rand.NextBool(5))
+                    {
+                        float angle = Main.rand.NextFloat(6.28f);
+                        Dust dust = Dust.NewDustPerfect((Projectile.Center - new Vector2(15, 15)) - (angle.ToRotationVector2() * (60 - decreasing)), ModContent.DustType<Dusts.NeedlerDustFive>());
+                        dust.scale = 0.05f;
+                        dust.velocity = angle.ToRotationVector2() * (Time < 60? 0.08f : 0.15f);
+                    }
+                    if (Projectile.timeLeft == 60)
+                    {
+                        SoundEngine.PlaySound(new SoundStyle($"{nameof(StarlightRiver)}/Sounds/Magic/FireCast"), Projectile.position);
+                    }
+                }
+                else if (Main.rand.NextBool(5) && !stuck)
+                {
+                    Dust.NewDustDirect(Projectile.position, Projectile.width, Projectile.height, ModContent.DustType<Dusts.GlassGravity>()).scale = Main.rand.NextFloat(0.8f, 1.1f);
+                }
+            }
+            switch (AmmoType)
```

Comment:
This hardcoded switch case seems far from ideal. It seems like it may be a better idea to abstract each projectile into its own wrapper class.